Edit textbox and upload pic

RankyLau opened the pull request:
Added layout and functionality for DescriptionPanel,
ImagePanel and InfoPanel, which is used to display the
data gathered from previous ViewModel.

RankyLau committed:
added transition from MainFragment to DescriptionPanelFragment, also set up the ui for DescriptionPanelFragment.

RankyLau committed:
created layout for ImagePanel, added method to add picture from the gallery or choose from the camera.
Also, added required permission to AndroidManifest

RankyLau committed:
created layout for ImagePanel, added method to add picture from the gallery or choose from the camera.
Also, added required permission to AndroidManifest

RankyLau committed:
created layout for ImagePanel, added method to add picture from the gallery or choose from the camera.
Also, added required permission to AndroidManifest

RankyLau committed:
created layout for ImagePanel, added method to add picture from the gallery or choose from the camera.
Also, added required permission to AndroidManifest

RankyLau committed:
Added field for image title

RankyLau committed:
Added field for image title

RankyLau committed:
Added field for image title

RankyLau committed:
Added field for image title

RankyLau committed:
Successfully used ViewModel to pass data around

RankyLau committed:
Successfully used ViewModel to pass data around

RankyLau committed:
Successfully used ViewModel to pass data around

RankyLau committed:
Successfully used ViewModel to pass data around

RankyLau committed:
Fixed image rotation problem

RankyLau committed:
Fixed image rotation problem

RankyLau committed:
Fixed image rotation problem

RankyLau committed:
Fixed image rotation problem

RankyLau committed:
Fixed issue with picture capture from camera

RankyLau committed:
Fixed issue with picture capture from camera

RankyLau committed:
Fixed issue with picture capture from camera

RankyLau committed:
Fixed issue with picture capture from camera

RankyLau committed:
Fixed issue with crashing when coming back from other activity
but with no result

RankyLau committed:
Fixed issue with crashing when coming back from other activity
but with no result

RankyLau committed:
Fixed issue with crashing when coming back from other activity
but with no result

RankyLau committed:
Fixed issue with crashing when coming back from other activity
but with no result

daphliu-zz commented on .idea/codeStyles/Project.xml:
```diff
@@ -0,0 +1,29 @@
+<component name="ProjectCodeStyleConfiguration">
```
Please delete the `/.idea/` folder and contents (**not** `/editor/.idea/`). Make sure you open up the editor folder in Android Studio, not the root soil-topArgraphy folder.

daphliu-zz commented on editor/app/build.gradle:
```diff
@@ -8,7 +8,7 @@ android {
     defaultConfig {
         applicationId "ca.ubc.eml.soiltopargraphy.editor"
-        minSdkVersion 15
+        minSdkVersion 24
```
Why did you need to change the minSdk? It would be ideal to support more phones if possible.

daphliu-zz commented on editor/app/build.gradle:
```diff
@@ -21,7 +21,10 @@ android {
     }
 }
-
+repositories {
```
I think you don't need to add these because they're already in [`/editor/build.gradle`](https://github.com/ubclaunchpad/soil-topARgraphy/blob/80544a9f5248115a8ee7c3af530b1bf33dd403c0/editor/build.gradle#L19)

daphliu-zz commented on editor/app/build.gradle:
```diff
@@ -32,4 +35,8 @@ dependencies {
     androidTestImplementation 'com.android.support.test.espresso:espresso-core:3.0.2'
+    implementation 'com.squareup.picasso:picasso:2.71828'
+    implementation 'com.github.bumptech.glide:glide:4.8.0'
+    annotationProcessor 'com.github.bumptech.glide:compiler:4.8.0'
```
I don't think we need both picasso and glide since they're similar libraries. Can you choose one of them?

daphliu-zz commented on editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/DescriptionPanelFragment.kt:
```diff
@@ -0,0 +1,57 @@
+
+class DescriptionPanelFragment : Fragment() {
+
+    private var mViewModel: DescriptionPanelViewModel? = null
```
```suggestion
    private lateinit var mViewModel: DescriptionPanelViewModel
```
Tip: you can use the `lateinit` keyword for variables that you set once, so you don't need the null checks later.

daphliu-zz commented on editor/app/src/main/AndroidManifest.xml:
```diff
@@ -9,13 +9,27 @@
+        </provider>
     </application>
-
+    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
```
Rather than do all the image loading in the app, I believe you can you Intents so that we don't need all these permissions for selecting an image from the phone camera or gallery. I found some [boilerplate code in this StackOverflow answer](https://stackoverflow.com/questions/10165302/dialog-to-pick-image-from-gallery-or-from-camera).

daphliu-zz commented on editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/InfoPanelFragment.kt:
```diff
@@ -0,0 +1,60 @@
+        val descriptionView = view.findViewById<TextView>(R.id.Info_panel_description)
+        val imageView = view.findViewById<ImageView>(R.id.ImageView)
+        val imageTitleView = view.findViewById<TextView>(R.id.Image_Title)
+        val imgmodel = activity?.run{
```
You may want to move this code to the `onActivityCreated` function on line 57, so that you can use the viewModel created there instead of re-creating it here.

daphliu-zz commented on editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/main/MainFragment.kt:
```diff
@@ -16,15 +19,27 @@ class MainFragment : Fragment() {
 
     private lateinit var viewModel: MainViewModel
 
+    fun onCreateInfoButtonClick(view: View){
```
```suggestion
    private fun onCreateInfoButtonClick(view: View){
```
nitpick: use a private function

daphliu-zz commented on editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/ImagePanelFragment.kt:
```diff
@@ -0,0 +1,202 @@
+
+    fun createImageFile(): File? {
+        val fileName = "MyPicture"
+        val storageDir = context?.getExternalFilesDir(Environment.DIRECTORY_PICTURES)
```
Could we use [internal storage instead of external storage](https://developer.android.com/training/data-storage/files#InternalVsExternalStorage)? That way we won't need the external storage permissions.

RankyLau committed:
Fixed issue with Glide,minSDK and repositories

RankyLau committed:
Fixed issue with Glide,minSDK and repositories

RankyLau committed:
Fixed issue with Glide,minSDK and repositories

RankyLau committed:
Fixed issue with Glide,minSDK and repositories

RankyLau committed:
Fixed issue about the following:
1. .idea folder
2. min sdk version
3. repositories
4. removed picasso from build.gradle
5. added lateint for some variables
6. removed requirement for reading and writing
to external storage, but not camera
7. changed most function to private function

RankyLau committed:
Fixed issue about the following:
1. .idea folder
2. min sdk version
3. repositories
4. removed picasso from build.gradle
5. added lateint for some variables
6. removed requirement for reading and writing
to external storage, but not camera
7. changed most function to private function

RankyLau committed:
Fixed issue about the following:
1. .idea folder
2. min sdk version
3. repositories
4. removed picasso from build.gradle
5. added lateint for some variables
6. removed requirement for reading and writing
to external storage, but not camera
7. changed most function to private function

RankyLau committed:
Fixed issue about the following:
1. .idea folder
2. min sdk version
3. repositories
4. removed picasso from build.gradle
5. added lateint for some variables
6. removed requirement for reading and writing
to external storage, but not camera
7. changed most function to private function

daphliu-zz commented on editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/DescriptionPanelFragment.kt:
```diff
@@ -0,0 +1,57 @@
+    private lateinit var mViewModel: DescriptionPanelViewModel
+    fun onToImageAddButtonClick(){
+        //record the field in ui to the viewModel
+        mViewModel?.name = view?.findViewById<EditText>(R.id.Title)?.text.toString()
```
`mViewModel` isn't nullable anymore so you can drop the `?.`
```suggestion
        mViewModel.name = view?.findViewById<EditText>(R.id.Title)?.text.toString()
```

ditto throughout

daphliu-zz commented on editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/ImagePanelViewModel.kt:
```diff
@@ -4,6 +4,6 @@ import android.arch.lifecycle.ViewModel
-    var uri: Uri?=null
-    var imageTitle: String?=null
+    lateinit var uri : Uri
+    lateinit var imageTitle: String
```
Don't go too overboard with `lateinit`, it makes sense for ViewModels since they are initialized once, but for properties inside the viewmodel that are set by the user they should be left as nullable.

RankyLau committed:
Removed some lateint in viewModel and some null check

RankyLau committed:
Removed some lateint in viewModel and some null check

daphliu-zz commented on editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/ImagePanelFragment.kt:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    private fun onToViewInfoPanelButtonClick() {
+        mViewModel?.imageTitle = view?.findViewById<EditText>(R.id.image_title)?.text.toString()
```
`mViewModel` isn't nullable anymore so you can drop the `?.`

daphliu-zz commented on editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/ImagePanelFragment.kt:
```diff
@@ -0,0 +1,171 @@
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        if (resultCode != RESULT_CANCELED) {
+            if (requestCode == CAMERA_PICK_CODE && resultCode == RESULT_OK) {
+                mViewModel?.uri = photoUri
```
ditto

daphliu-zz commented on editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/ImagePanelFragment.kt:
```diff
@@ -0,0 +1,171 @@
+        if (resultCode != RESULT_CANCELED) {
+            if (requestCode == CAMERA_PICK_CODE && resultCode == RESULT_OK) {
+                mViewModel?.uri = photoUri
+                mImageView.setImageURI(mViewModel?.uri)
```
ditto

daphliu-zz reviewed (approved):
Looks good! There's a couple other spots in `ImagePanelFragment` to drop the `?.` after `mViewModel`. Once those are fixed we can merge this branch!

RankyLau committed:
rebased with the master branch, and added button to save info panel, right now I have removed quizPanel assiocation with it
since it has not been implemented in the master branch

RankyLau committed:
rebased with the master branch, and added button to save info panel, right now I have removed quizPanel assiocation with it
since it has not been implemented in the master branch

RankyLau committed:
Merge remote-tracking branch 'origin/edit_textbox_and_upload_pic' into edit_textbox_and_upload_pic

# Conflicts:
#	editor/app/build.gradle
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/flag/FlagMapFragment.java
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/DescriptionPanelFragment.kt
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/DescriptionPanelViewModel.kt
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/ImagePanelViewModel.kt
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/InfoPanel.kt
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/InfoPanelFragment.kt
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/InfoPanelViewModel.kt
#	editor/app/src/main/res/layout/info_panel_fragment.xml
#	editor/app/src/main/res/values/strings.xml

daphliu-zz commented:
I think after the rebase, a lot of your changes from comments were undone. I tried to mark all the fixed comments above as resolved, and left the ones that still apply to the new code.

RankyLau committed:
Fixed all of the issues mentioned

RankyLau committed:
Apprantely I somehow didn't import the converter file during
my rebase, so this should have included the converter file after
another rebase

RankyLau committed:
Merge remote-tracking branch 'origin/edit_textbox_and_upload_pic' into edit_textbox_and_upload_pic

# Conflicts:
#	editor/app/src/main/AndroidManifest.xml
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/DescriptionPanelFragment.kt
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/DescriptionPanelViewModel.kt
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/ImagePanelFragment.kt
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/ImagePanelViewModel.kt
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/InfoPanelFragment.kt
#	editor/app/src/main/java/ca/ubc/eml/soiltopargraphy/editor/ui/infopanel/InfoPanelViewModel.kt
#	editor/app/src/main/res/layout/main_fragment.xml

RankyLau committed:
edited back minsdk 15

daphliu-zz commented on editor/app/src/main/res/layout/image_panel_fragment.xml:
```diff
@@ -1,13 +1,92 @@
+        android:id="@+id/add_image_via_gallery_button"
+        android:layout_width="0dp"
+        android:layout_height="63dp"
+        android:layout_marginStart="8dp"
```
To fix the build issues, can you add or replace this with `android:layout_marginLeft`? Similarily, we need to use `android:layout_marginRight` for `marginEnd`.
```diff
        android:layout_marginStart="8dp"
+       android:layout_marginLeft="8dp"
        android:layout_marginTop="35dp"
        android:layout_marginEnd="8dp"
+       android:layout_marginRight="8dp"
        android:layout_marginBottom="15dp"
```

daphliu-zz commented on editor/app/src/main/res/layout/image_panel_fragment.xml:
```diff
@@ -1,13 +1,92 @@
+        android:id="@+id/add_image_via_camera_button"
+        android:layout_width="0dp"
+        android:layout_height="63dp"
+        android:layout_marginStart="8dp"
```
margin again

daphliu-zz commented on editor/app/src/main/res/layout/image_panel_fragment.xml:
```diff
@@ -1,13 +1,92 @@
+        android:id="@+id/to_view_info_panel_button"
+        android:layout_width="0dp"
+        android:layout_height="63dp"
+        android:layout_marginStart="50dp"
```
margin again

daphliu-zz commented on editor/app/src/main/res/layout/image_panel_fragment.xml:
```diff
@@ -1,13 +1,92 @@
+        android:id="@+id/image_title"
+        android:layout_width="0dp"
+        android:layout_height="0dp"
+        android:layout_marginStart="8dp"
```
margin again

daphliu-zz commented on editor/app/src/main/res/layout/info_panel_fragment.xml:
```diff
@@ -0,0 +1,93 @@
+        android:id="@+id/save_info_panel"
+        android:layout_width="290dp"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="8dp"
```
margin again

daphliu-zz reviewed (changes requested):
Great! The external storage code is still lingering, and Travis is failing due to a lint issue for older Android versions. Once those are updated, we can merge this in.

RankyLau committed:
Added method to write infoPnael as JSON,also added support for sdk 15

RankyLau committed:
Merge remote-tracking branch 'origin/edit_textbox_and_upload_pic' into edit_textbox_and_upload_pic

daphliu-zz committed:
added marginleft and marginright

RankyLau committed:
removed read_external_storage

RankyLau committed:
Merge branch 'master' into edit_textbox_and_upload_pic

RankyLau committed:
Fixed issue with travis
